### PR TITLE
feat: support phi3.5 moe

### PR DESCRIPTION
**Pull request by drbh**
This is a work in progress PR to add support for [microsoft/Phi-3.5-MoE-instruct](https://huggingface.co/microsoft/Phi-3.5-MoE-instruct)

TODO
- [X] add phi 3.5 to `ModelType`
- [X] load weights into memory
- [X] prefer moe over mlp in layers
- [X] enable long/short rope scaling
- [x] validate scaling logic
- [x] ensure layer logic is correct
- [x] ensure no regressions on existing phi models
- [x] identify issue with allocating graphs
- [x] refactor/cleanup/add tests

### CODE REVIEW

File: integration-tests/models/test_flash_phi35_moe.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                "role": "user",
+                "content": "What is the weather like in Brooklyn, New York?",
+            },
+        ],
+    )
+
+    print(repr(response.choices[0].message.content))
+    assert (
+        response.choices[0].message.content
+        == "I'keeper services don't have real-time capabilities, however, I can guide you on how to find current weather conditions in Brooklyn, New York.\n\nTo get the most accurate"
```
Comment (drbh) [inline integration-tests/models/test_flash_phi35_moe.py:38]: - [ ] address `I'keeper` issue (likely related to tokenizer) before ready to merge

**Comment by drbh** [comment]
This PR now adds phi 3.5 moe, and improves the chat endpoint to assume greedy generation unless the temp is explicitly set  by the user in the request (this helped align the expected output from phi with the reference impl).

Start phi3.5moe
```bash
text-generation-launcher \
  --model-id microsoft/Phi-3.5-MoE-instruct \
  --num-shard 4 \
  --cuda-graphs 1,2 \
  --trust-remote-code
```

send a request
```bash
curl 127.0.0.1:3000/generate -X POST \
  -H 'Content-Type: application/json' \
  -d '{
    "inputs": "Hello who are you?",
    "parameters": {
      "max_new_tokens": 20
    }
  }'
```

response

```json
{
    "generated_text": " I'm an artificial intelligence developed by Microsoft to assist with a variety of tasks and provide information."
}
```

File: server/text_generation_server/layers/rotary.py [hunk 2]
```diff
@@ -89,6 +89,43 @@ def static(cls, config, dim, base, device):
 
             if rope_type == "linear":
                 pass
+            elif rope_type == "longrope":
```
Comment (danieldk) [inline server/text_generation_server/layers/rotary.py:92]: Discussed in Slack: there is also another `longrope` condition later in this conditional block.

Comment (drbh) [reply server/text_generation_server/layers/rotary.py]: thanks for pointing this out 🙏, i've consolidate the two longropes paths in the latest commit and only use the new class if `"short_mscale"` and `"short_mscale"` are in the config

**Review by Narsil — DISMISSED** [review]

**Review by Narsil — APPROVED** [review]

**Pull request merged**